Indexing permissions as part of the Elastic Agent policy

aleksmaus opened the pull request:
## What does this PR do?

Implements the Indexing permissions as part of the Elastic Agent policy
https://github.com/elastic/fleet-server/issues/101

1. Adds a new field ```policy_output_permissions_hash``` into the Agent document that stores the hash of ```output_permissions``` from the policy.
2. Retrieves the policy from .fleet-policies upon the agent enrollment and uses the ```output_permissions``` ```default``` namespace in order to create the role descriptors for the output API key generation.
3. Checks the ```output_permissions``` upon every checking. If the hash doesn't match with the stored in the agent document the new key is generated.

This PR:
1. doesn't invalidate the old API key see the comment for the next iteration https://github.com/elastic/fleet-server/issues/101#issuecomment-806286176
2. is not fully validated end-to-end with kibana for the ```output_permissions``` changes, pending on further kibana changes.

Should only be merged after the corresponding kibana changes:
https://github.com/elastic/kibana/pull/94591


## Why is it important?

Addresses: https://github.com/elastic/fleet-server/issues/101

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation


## Related issues

- Addresses https://github.com/elastic/fleet-server/issues/101

## Screenshots

<img width="973" alt="Screen Shot 2021-03-24 at 6 40 40 PM" src="https://user-images.githubusercontent.com/872351/112406932-874b6300-8ceb-11eb-91cd-075c27b4d7fe.png">

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Ffleet-server%2FPR-169/detail/PR-169/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Ffleet-server%2FPR-169/detail/PR-169/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Ffleet-server%2FPR-169/detail/PR-169/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Ffleet-server%2FPR-169/detail/PR-169/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://fleet-server_169.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #169 opened`



* Start Time: 2021-03-25T01:57:51.258+0000


* Duration: 5 min 49 sec


* Commit: 2e87710e085e55c79c41a3bc2121393c425afe49



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 66  |
| Skipped      | 0 |
| Total        | 66   |


#### Trends :test_tube:


![Image of Build Times](https://beats-ci.elastic.co/job/Ingest-manager/job/fleet-server/job/PR-169/buildTimeGraph/png)


![Image of Tests](https://beats-ci.elastic.co/job/Ingest-manager/job/fleet-server/job/PR-169/1/testReport/history/countGraph/png?start=0&amp;end=25)


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












<!--COMMENT_GENERATED_WITH_ID_comment.id-->

ruflin commented on internal/pkg/policy/output_permissions.go:
```diff
@@ -0,0 +1,160 @@
+		}
+
+		m := make(smap.Map)
+		m["index"] = idx
```
I hope we can get rid of this bit as it is planned that Kibana will add the ` index` part to the output so we don't need to do the additional processing here.

aleksmaus replied:
Yeah, Alejandro reached out about the upcoming change. Might add more things here then.

blakerouse commented on internal/pkg/dl/policies.go:
```diff
@@ -79,3 +90,18 @@ func CreatePolicy(ctx context.Context, bulker bulk.Bulk, policy model.Policy, op
 }
+
+// FindPolicyByID find policy by ID
+func FindPolicyByID(ctx context.Context, bulker bulk.Bulk, policyID string) (policy model.Policy, err error) {
```
The `policy.Monitor` keeps the latest revision of each policy in memory. It would be better to fetch the policy from there instead of going to elasticsearch, as fleet-server already has it cached in memory.

aleksmaus replied:
Sounds good, was not aware, will update.

blakerouse reviewed (approved):
Overall this change looks good. See my inline comment about fetching the policy from elasticsearch.

aleksmaus commented:
Removed the output key generation upon enrollment. Not this is driven by the policy update pushes only, one code path.
Updated to the latest kibana changes, removed output permissions to the role descriptors transformation.

ruflin commented:
@aleksmaus I see this got merged but https://github.com/elastic/kibana/pull/94591 is still open. How is the current behaviour?
